spatialdb/object_indices.py

```diff
@@ -655,3 +725,355 @@ def reserve_ids(self, resource, num_ids, version=0):
+
+        Returns:
+            (bool, int): True if cuboid found and the chunk_num of the key that contains the cuboid.  False, -1 if cuboid isn't found.
+        """
+        for i in range(last_chunk_num + 1):
+            if i == 0:
+                key = default_key
+            else:
+                key = '{}&{}'.format(default_key, i)
+
+            # Using a query so we don't actually return the set to the client.
+            resp = self.dynamodb.query(
+                TableName=self.id_index_table,
+                Select='COUNT',
+                KeyConditionExpression='#key=:keyvalue AND version=:version',
+                ExpressionAttributeNames={
+                    '#key': 'channel-id-key',
+                    '#cuboidset': 'cuboid-set'},
+                ExpressionAttributeValues={
+                    ':keyvalue': {'S': key},
+                    ':version': {'N': str(version)},
+                    ':morton': {'S': str(cuboid_morton)}},
+                FilterExpression='contains(#cuboidset, :morton)')
+
+            # If result set non empty, return True
+            if resp['Count'] > 0:
+                return (True, i)
+
+        return (False, -1)
+
+    def write_cuboid(
+            self, max_used_capacity, cuboid_morton, key, chunk_num, rev_id, 
+            lookup_key, version=0):
+        """
+        Write the morton id to the given key.
+
+        Will increment the key's chunk_num as necessary if the given key and chunk_num
+        combination's set is full or max_used_capacity is exceeded.
+
+        Args:
+            max_used_capacity (int): After updating the table, if used write capacity exceeded this value, update the LAST_PARTITION_KEY for this object id.
+            cuboid_morton (string|int): Morton id of the cuboid.
+            key (str): Key in the id index table without a chunk_num.
+            chunk_num (int): Number to be appended to the key if greater than 0.
+            rev_id (int|None): Expected revision id when updating cuboid-set attribute.
+            lookup_key (str): Key that identifies the channel and resolution of the cuboid.
+            version (optional[int]): Version - reserved for future use.
+
+        Returns:
+            (int): chunk that the morton id was written to.
+        """
+        new_chunk_num = chunk_num
+        exp_rev_id = rev_id
+
+        done = False
+
+        while not done:
+            if new_chunk_num > 0:
+                dynamo_key = '{}&{}'.format(key, new_chunk_num)
+            else:
+                dynamo_key = key
+
+            try:
+                response = self.write_cuboid_dynamo(
+                    cuboid_morton, dynamo_key, exp_rev_id, lookup_key, version)
+                if 'ConsumedCapacity' in response:
+                    used = response['ConsumedCapacity']['CapacityUnits']
+                    if used >= max_used_capacity:
+                        new_chunk_num = new_chunk_num + 1
```

Comment:
I'm slightly concerned with slightly changing the key to solve write capacity issues (completely). I get that the key gets much smaller, so needs less capacity, but the same partition can still be getting hammered, right? Or is the team pretty confident from empirical testing that it's fine like this, and really just dealing with keys that blow past 400k is enough? 

Since indexing now runs out of the path of ingest, possibly this is all fine as you can better control throughput?

Reply:
Unfortunately, the real empirical testing won't happen until I get to try indexing some of the phase 1 datasets. :) And of course that still won't necessarily reflect what the phase 2 datasets will look like. 

Currently, each cuboid that contains the same id will try to add itself to that id in a separate lambda/step fcn and only one can update the row at a time. They'll each have to wait their turn. So I think a single partition won't be hit too hard once the step fcns start backing off.